docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+# User-Defined Type Functions
+
+## Summary
+
+The new type solver for Luau supports a limited number of built-in type functions that enable developers to be expressive with their code. To further enhance this expressivity and the flexibility of the type system, this RFC proposes the design and implementation of user-defined type functions, a feature that will allow developers to define custom type functions.
+
+## Motivation
+
+The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
+- Enable more precise type definitions that can capture complex relationships and constraints
+- Facilitate the creation of reusable and composable libraries
+- Enhance code safety and maintainability with better type-level programming
+
+The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.
+
+## Design
+
+Type functions wil be defined with the following syntax:
+```luau
+type function f(...)
+    -- implementation of the type function
+end
+```
+
+For instance, the `rawget` type function can be written as:
+```luau
+type function rawget(tbl, prop)
+    return tbl.getProps()[prop.getValue()]
+
+    -- returning nothing / nil means that this type function will fail to reduce
+end
+
+type Person = {
+    name: string,
+    age: number
+}
+
+type ty = rawget<Person, "name"> -- ty = string
+```
+To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `lType`.  An `lType` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `lType`. Moreover, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
```

Comment:
I don't think we should actually be calling the userdata `lType`. Certainly, the _header_ has that name because that's consistent with how we name Luau VM headers (a convention adopted from Lua itself), but the user-facing name should just be `type` or `types`. `type` would be most consistent with `string`, `table,` buffer`, but it's a bit thorny because it would mean shadowing the global function `type` that is already present in the VM. Possibly a better option is `luautype`, or maybe someone will have alternative suggestions for the library name.

Reply:
I should've mentioned that lType was a placeholder for the name in the RFC. I was also concerned with just calling it `type`. I think `luautype` is a good name.

Reply:
Having luautype is strange for this being inside the Luau language.

Reply:
What about a variation of `metatype`?

Reply:
we've decided to go with `typelib`!